official-antistasi-community/A3-Antistasi
# CS Dismissible Hints

![IE_cannot_display_webpage](https://user-images.githubusercontent.com/31820984/90346507-be58e800-e029-11ea-96bc-d94f51d1f73a.jpg)
## What type of PR is this.
* Bug
* Change

## What have you changed and why?
Information:
Notifications can now be dismissed.
`customHint` has been split into `customHint` and `renderHint`. (renderHint is not called directly)
Notifications can be longer than 8kb as customHint uses joinString rather than format.
The new system runs between 3x, 6x, 15x faster than the original implementation.
Full backwards compatibility

### Specifics
Notifications are queued up in a player's global list. (`A3A_NotifQueue`) 
The dismiss key is bound to use action 12. (Will be dynamic when clientSide preferences are a thing)
The dismiss key or "Use Action 12" is displayed in a notifications footer.
The icon can be changed by specifying an image path with an aspect ratio.
Notifications can be pre-processed, allowing control over the entire message except for the footer.
New notifications will overwrite old notifications with a matching header.

## Please verify the following and ensure all checks are completed.

2. [x] Have you loaded the mission in LAN host?
3. [x] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
* Yes (Further Testing would be appreciated.)

### How can the changes be tested?
> Basic
Make sure mission boots.
Boot messages overwrite each other without a dismiss option.
At the end of init, the dismiss option is enabled.
Create some messages and cycle through them on a low end PC.

> Run the examples:
```sqf
["FooBar", "Hello World"] call A3A_fnc_customHint;

//Save Notify
        ["Restart Notification", "Please make your way to HQ as a restart will occure soon™<br/><br/>Please make sure you save your stats at the Map.", true] remoteExec ["A3A_fnc_customHint", 0, false];

["Vaya...", "Parece que sus notificaciones importantes se cifraron.<br/><br/>Nadie espera el cifrado español.", false, ["Pictures\Intel\laptop_error.paa",1]] remoteExec ["A3A_fnc_customHint", 0, false];

// Pre-parse FooBar(Hello World) NoMacro
        private _iconXML = parseText "<img color='#ffffff' image='functions\UI\images\logo.paa' align='center' size='2' shadow='1' shadowColor='#000000' />";
        private _separator  = parseText "<br/><img color='#e6b24a' image='functions\UI\images\img_line_ca.paa' align='center' size='0.60' />";
        private _header = parseText "<br/><br/><t size='1.2' color='#e5b348' shadow='1' shadowColor='#000000'>FooBar</t>";
        private _body = parseText "<br/><br/><t size='1' color='#ffffff' shadow='1' shadowColor='#000000'>Hello World</t><br/>";
        FooBarParse = composeText [_iconXML, _header, _separator, _body, _separator];
        ["FooBar", FooBarParse] call A3A_fnc_customHint;

```
> Push your PC's hardware.
https://github.com/CalebSerafin/FrostsBites-Random-Arma-Scripts/blob/master/A3A_HintRainbowUberFHD.sqf
********************************************************
There will be a follow-up PR to remove any `hintSilent ""` as there is a number that hides the save screen for the first <10 secs.
There will be a follow-up PR to completely replace vanilla hint with a control display.

Familiar Unit traits notification:
![Unit_traits_hint](https://user-images.githubusercontent.com/31820984/90346800-6079cf80-e02c-11ea-833c-6a048281e2aa.jpg)
